node/pkg/common/queryResponse.go

```diff
@@ -206,103 +149,182 @@ func UnmarshalQueryResponsePublication(data []byte) (*QueryResponsePublication,
 
-	return msg, nil
+	return nil
 }
 
-func UnmarshalQueryPerChainResponseFromReader(reader *bytes.Reader) (*PerChainQueryResponse, error) {
-	pcr := PerChainQueryResponse{}
+// Validate does basic validation on a received query request.
+func (msg *QueryResponsePublication) Validate() error {
+	// Unmarshal and validate the contained query request.
+	var queryRequest QueryRequest
+	err := queryRequest.Unmarshal(msg.Request.QueryRequest)
+	if err != nil {
+		return fmt.Errorf("failed to unmarshal query request")
+	}
+	if err := queryRequest.Validate(); err != nil {
+		return fmt.Errorf("query request is invalid: %w", err)
+	}
+
+	if len(msg.PerChainResponses) == 0 {
+		return fmt.Errorf("response does not contain any per chain responses")
+	}
+	if len(msg.PerChainResponses) > math.MaxUint8 {
+		return fmt.Errorf("too many per chain responses")
+	}
+	for idx, pcr := range msg.PerChainResponses {
+		if err := pcr.Validate(); err != nil {
+			return fmt.Errorf("failed to validate per chain query %d: %w", idx, err)
+		}
+	}
+	return nil
+}
 
-	chainID := uint32(0)
-	if err := binary.Read(reader, binary.BigEndian, &chainID); err != nil {
-		return nil, fmt.Errorf("failed to read chain ID: %w", err)
+// Equal checks for equality on two query response publications.
+func (left *QueryResponsePublication) Equal(right *QueryResponsePublication) bool {
+	if !bytes.Equal(left.Request.QueryRequest, right.Request.QueryRequest) || !bytes.Equal(left.Request.Signature, right.Request.Signature) {
+		return false
+	}
+	if len(left.PerChainResponses) != len(right.PerChainResponses) {
+		return false
+	}
+	for idx := range left.PerChainResponses {
+		if !left.PerChainResponses[idx].Equal(&right.PerChainResponses[idx]) {
+			return false
+		}
+	}
+	return true
+}
+
+func (resp *QueryResponsePublication) RequestID() string {
+	if resp == nil || resp.Request == nil {
+		return "nil"
+	}
+	return hex.EncodeToString(resp.Request.Signature)
+}
+
+// Similar to sdk/vaa/structs.go,
+// In order to save space in the solana signature verification instruction, we hash twice so we only need to pass in
+// the first hash (32 bytes) vs the full body data.
+// TODO: confirm if this works / is worthwhile.
+func (msg *QueryResponsePublication) SigningDigest() (common.Hash, error) {
+	msgBytes, err := msg.Marshal()
+	if err != nil {
+		return common.Hash{}, err
+	}
+	return GetQueryResponseDigestFromBytes(msgBytes), nil
+}
+
+// GetQueryResponseDigestFromBytes computes the digest bytes for a query response byte array.
+func GetQueryResponseDigestFromBytes(b []byte) common.Hash {
+	return crypto.Keccak256Hash(append(queryResponsePrefix, crypto.Keccak256Hash(b).Bytes()...))
+}
+
+//
+// Implementation of PerChainQueryResponse.
+//
+
+// Marshal marshalls a per chain query response.
+func (pcr *PerChainQueryResponse) Marshal() ([]byte, error) {
+	if err := pcr.Validate(); err != nil {
+		return nil, err
+	}
+
+	buf := new(bytes.Buffer)
+	vaa.MustWrite(buf, binary.BigEndian, pcr.ChainID)
```

Comment:
If it's possible that ChainID isn't unique in a request (which seems reasonable), then should we avoid writing it here if this response list should match up 1-1 with the request list?

Reply:
We can play with this later 🙂